library/Hasql/Pool.hs

```diff
@@ -34,10 +124,33 @@ data Pool = Pool
     poolReuse :: TVar (TVar Bool)
   }
 
+data Pool = Pool
+  { pool :: RawPool,
+    reaperRef :: IORef ()
+  }
+
 -- | Create a connection-pool.
 --
 -- No connections actually get established by this function. It is delegated
 -- to 'use'.
+acquireConf :: Config -> IO Pool
+acquireConf config = do
+  rawPool <-
+    RawPool config
+      <$> newTQueueIO
+      <*> newTVarIO (confSize config)
+      <*> (newTVarIO =<< newTVarIO True)
+  ref <- newIORef ()
+  manager <- forkIOWithUnmask $ \unmask -> unmask $ manage rawPool
+  void . mkWeakIORef ref $ do
+    -- When the pool goes out of scope, stop the manager.
```

Comment:
I've verified this works via the following change. Seems a bit too awkward to test automatically, what do you think?

```
diff --git a/library/Hasql/Pool.hs b/library/Hasql/Pool.hs
index 1a3d0b9..17d17d7 100644
--- a/library/Hasql/Pool.hs
+++ b/library/Hasql/Pool.hs
@@ -141,9 +141,10 @@ acquireConf config = do
       <*> newTVarIO (confSize config)
       <*> (newTVarIO =<< newTVarIO True)
   ref <- newIORef ()
-  manager <- forkIOWithUnmask $ \unmask -> unmask $ manage rawPool
+  manager <- forkIOWithUnmask $ \unmask -> unmask $ (putStrLn "forking" >> manage rawPool)
   void . mkWeakIORef ref $ do
     -- When the pool goes out of scope, stop the manager.
+    putStrLn "killing manager"
     killThread manager
   return $ Pool rawPool ref
 
diff --git a/test/Main.hs b/test/Main.hs
index 9ddf141..f8e91cb 100644
--- a/test/Main.hs
+++ b/test/Main.hs
@@ -10,6 +10,7 @@ import Hasql.Pool
 import qualified Hasql.Session as Session
 import qualified Hasql.Statement as Statement
 import qualified System.Environment
+import System.Mem (performGC)
 import qualified System.Random as Random
 import qualified System.Random.Stateful as Random
 import Test.Hspec
@@ -120,6 +121,8 @@ main = do
               res3 <- use countPool $ countConnectionsSession appName
               res3 `shouldBe` Right 0
           )
+      performGC
+      threadDelay 5000000 -- 1s
 
 getConnectionSettings :: IO Connection.Settings
 getConnectionSettings =
```

Reply:
The whole weak ref thing is lifted more or less directly from [resource-pool](https://github.com/scrive/pool/blob/0a05714739fc533c9840a8a5cb8bdd07ead57800/src/Data/Pool/Internal.hs#L143). I'm not 100% on the exception handling part here. I *think* if the manager action throws it'll spam stderr, but also it shouldn't throw because it's just closing connections, but... bit of room for things to go wrong here.

Reply:
This can see how this could be tested if we were to extract a general (not hasql-specialized) lib from here. But until we do I guess it's okay :)